### ISSUE TITLE
TF24 hydraulics: Performance improvement for spline class

### ISSUE THREAD

**Issue by hol430** [traitecoevo/plant#435]
Hi all, we've adapted your approach for using splines to interpolate water potential from transpiration and vice versa in the lpj-guess dave model. This has been an excellent optimisation for us, so thanks for sharing the approach! One thing that we noticed however, was that a lot of time was spent in `std::lower_bound()` as called from [`spline::operator ()`](https://github.com/traitecoevo/plant/blob/develop/src/tk_spline.cpp#L207). This is called a lot in our model, and 20% of the total runtime was spent inside this particular call to `std::lower_bound()`.

What's happening here is we're searching through the x-values to find a start point for the interpolation. `std::lower_bound()` is doing a binary search internally, so it's reasonably efficient, but what we did to try and speed things up was to do a one-time check (when we initialise the spline object) to check if the x-values are equally-spaced. If they are, then we can eliminate the call to `std::lower_bound()` at time-of-interpolation by looking at our x value relative to the full range of x values (ie `(x - xmin) / (xmax - xmin)`). We then multiply that by the number of x-values and floor() the result, and end up with something comparable to the output of `std::lower_bound()` in significantly less time than it would have otherwise taken.

After doing that (and ensuring we chose equidistant x-values), we managed to speed up our simulations by 18%, so I thought I'd share our findings here in case it's of use to you guys. I get that your model works differently, so if this is not really an issue for you guys (e.g. if this code is not called so often), then feel free to close this issue :).

Edit: didn't mean to label this as a bug, but I couldn't see a better category.

**Comment by dfalster** [comment]
Apologigies for the slow response @hol430. Brilliant! very glad the interpolation helped you speed up your code. And great idea about the additional optimisation. We'll implement that also.

**Comment by dfalster** [comment]
Thanks @hol430 — we implemented this and it's a clear win. Reporting back in case it's useful.

We're working on a new TF24 plant-hydraulics model (on our `root_water_uptake` branch) whose solver hammers these splines extremely hard: for every individual at every ODE sub-step it solves the soil→root→stem→leaf continuum via nested 1-D root-finds plus a golden-section profit optimisation, each iteration evaluating the transpiration / root-vulnerability splines many times. Profiling confirmed your finding — `std::lower_bound` inside `spline::operator()` was the single hottest leaf.

**Result:** on a representative run this change (combined with a small memoisation elsewhere) cut runtime roughly in half (~41s → ~21s, ≈1.9×), and the trajectory is **bit-identical** to before (`all.equal` TRUE). The gain is larger than your ~18% simply because our solver leans on the splines so much harder.

**Implementation** (all inside `tk::spline`, so every call site benefits automatically):

- In `set_points()`, after the coefficients are built, we do a one-time check for an equidistant x-grid and cache `m_uniform`, `m_x0` and `m_inv_dx = 1/dx` (tolerance `1e-6 * dx`).
- In `operator()`, when `m_uniform` we compute the index directly as `idx = (int)((x - m_x0) * m_inv_dx)` instead of `std::lower_bound`.

One wrinkle worth flagging: a plain `floor((x - xmin)/(xmax - xmin) * n)` is *comparable* to `lower_bound` but not guaranteed identical at knot-rounding boundaries, and because our results feed an adaptive ODE solver, even a 1-index difference at a knot can perturb the whole trajectory. To stay bit-identical we follow the direct index with a tiny correction that reproduces `lower_bound` exactly (runs 0 times for essentially all queries):

```cpp
idx = static_cast<int>((x - m_x0) * m_inv_dx);
const int idx_max = static_cast<int>(n) - 2;
if (idx < 0) idx = 0; else if (idx > idx_max) idx = idx_max;
while (idx > 0       && m_x[idx]   >= x) --idx;   // ensure m_x[idx] < x
while (idx < idx_max && m_x[idx+1] <  x) ++idx;   // ensure x <= m_x[idx+1]
```

The non-uniform path is untouched (`else` branch still calls `std::lower_bound`), so there's no regression for adaptively-refined splines.

**What's needed to apply it elsewhere:** nothing at the call sites. Because the optimisation lives in `tk::spline` and is gated on the auto-detected `m_uniform` flag, *every* spline built on an evenly-spaced x-grid speeds up transparently, and anything non-uniform falls back to the binary search unchanged. The only "action" to extend the benefit to a spline that doesn't currently see it is to construct it on an equidistant x-grid. In `plant` all splines go through `Interpolator` → `tk::spline::set_points`, so the leaf transpiration / root-vulnerability splines (built with a constant step) benefit immediately, while the adaptively-refined canopy/light splines correctly stay on the fallback path.

This is currently on our `root_water_uptake` branch; the `tk::spline` change is self-contained and could be cherry-picked to `develop` independently of the hydraulics work. Happy to open a PR for just that piece if useful.

**Comment by hol430** [comment]
Oh that's brilliant - 2x speedup is crazy for such an easy win. Glad to hear it helped!

**Comment by dfalster** [comment]
Yes crazy returns! We first made the model really slow and  hitting the spline over and over. So it was a somewhat easy but very valuable win.

**Comment by dfalster** [comment]
I previously reported big gains in the hydraulic component. I've now also tested how this improves use of the spline in a light environment, and here gains are more modest.

For the FF16 light-availability spline, porting the O(1) uniform-grid spline index gave a small but consistent speedup: about 1.02-1.03x in `run_plant_benchmarks()` (`scm`: 3.965 s -> 3.870 s over 20 repeats; `build_schedule`: 10.825 s -> 10.470 s).

The likely reason is that the light spline is often adaptive rather than uniformly spaced. In the collected FF16 `scm` run, 84/142 light-spline snapshots were uniform, but 58/142 were non-uniform; the final spline had 101 knots with 5 distinct gap sizes. For the schedule-refinement case, 94/180 snapshots were uniform and 86/180 non-uniform. So the uniform-grid fast path helps part of the FF16 light run, but not the later adaptive spline states.

This contrasts with the hydraulic case, where the relevant spline lookups appear to spend much more time on an equidistant grid, so replacing `std::lower_bound()` has a much larger effect.

**Comment by dfalster** [comment]
closed via https://github.com/traitecoevo/plant/pull/488 and https://github.com/traitecoevo/plant/pull/471
